Auth doc refactor (#5598)
strongloop/loopback-next: docs/site/Authentication-overview.md

---
lang: en
title: 'LoopBack 4 Authentication Overview'
keywords: LoopBack 4.0, LoopBack 4, Authentication
sidebar: lb4_sidebar
permalink: /doc/en/lb4/Authentication-overview.html
---

Security is of paramount importance when developing a web or mobile application
and usually consists of two distinct pieces:

- Authentication
- Authorization

Authentication is a process of verifying someone's identity before a protected

> resource is accessed.

Reviewer: Authentication is the process of validating the user/entity to the system, which enables authorised access to the protected routes.

> Authentication is a process of verifying someone's identity before a protected
> resource is accessed.
> 

Reviewer: ```suggestion
Authentication is a process of verifying user/entity to the system, which enables identified/validated access to the protected routes.
```

Authorization is a process of deciding if a user can perform an action on a
protected resource.

{% include note.html content=" For a description of an Authorization process, please see [Authorization](Loopback-component-authorization.md). " %}

This document gives you an overview of the authentication system provided in
LoopBack 4.

> Let's start with the following scenario: Suppose you want to secure endpoint
> `GET /todo` using
> [JWT(JSON web token)](https://github.com/auth0/node-jsonwebtoken#readme)
> strategy, the diagram below shows how such authentication process works with
> LoopBack's authentication mechanism.

Reviewer 2: ```suggestion
Let's start with the following scenario: Suppose you want to secure endpoint
`GET /todo` using a
[JWT (JSON Web Token)](https://github.com/auth0/node-jsonwebtoken#readme)
strategy. The diagram below shows how such authentication process works with
LoopBack's authentication mechanism.
```

Minor rewording for clarity.

> ![authentication_overview_request_handle_flow](./imgs/authentication/authentication-overview.png)

Reviewer 3: there're extra icons at the bottom right. Could you please remove that?

Author: The diagram is created kinda in a rush, I will draw it again.

Author: @dhmlau I updated the diagram.

> As you can see, a request with token in its header reaches the application. It's
> handled by the REST server's sequence, which invokes the authentication action
> to decode the user profile from token so that controller can inject the user
> into its functions. All you need to add is the code in programming colors:

Reviewer: ```suggestion
As illustrated above, a access token in Authorization header in making request to the route is 
handled by the REST server's sequence, which invokes the authentication action
to decode the user profile from token so that controller can inject the user to the system and perform the actions. For implementing this, All you need: is to add the code in programming colors:
```

Author: @madaky 👍 The controller doesn't execute action, I rephrased it to 

> As illustrated above, an access token in Authorization header in making request to the route is handled by the REST server's sequence, which invokes the authentication action to decode the user profile from token so that controllers can have the user injected.

> For implementing this, all you need to add is the code in programming colors:

- Register the authentication component and JWT extension in your application.
- Enable the authenticate action in sequence.
- Decorate the controller endpoint with `@authenticate()` and inject the user
  passed from the authentication layer.

The rest will be handled by our authentication component

> `@loopback/authentication`, which glues all the parts together, and the JWT
> extension `@loopback/jwt-authentication`, which is usually provided by extension
> developers.

Reviewer: ```suggestion
`@loopback/authentication`, which incorporates the authentication mechanism, and the JWT
extension `@loopback/jwt-authentication`, which helps in implementing jwt-based authentication to the system,  provided by extension developers.
```

> The authentication system is highly extensible and plugible. It's easy to get
> started with. While there are more advanced features to explorer, and they
> benefit when you build more complicated and larger scale applications. This
> requires you to understand the mechanism of the system, like how component
> `@loopback/authentication` works. To help you learn the full features gradually,
> we break down the documentations into several parts:

Reviewer 2: ```suggestion
The authentication system is highly extensible and pluggable. It's easy to get
started with. While there are more advanced features to explore - of which
are beneficial when you build more complicated and larger scale applications -
, it requires understanding the mechanics of the system, like how
`@loopback/authentication` works. To help you learn the full features gradually,
we've broken down the documentations into several parts:
```

Fixed typos and some rewording to be less direct (e.g. remove reference to the reader with words such as "you")

> - [JWT todo example](missing_link): A tutorial to get started by applying the

Reviewer 4: I think for these links, we can highlight the JWT todo link to show that this is the first step to start with for those who are new to this area.

  JWT authentication in the
  [todo example](https://loopback.io/doc/en/lb4/todo-tutorial.html).
- Understand the modules provided out-of-the-box:
  - [Authentication component](missing_link): A deep dive of component
    `@loopback/authentication`
  - [JWT extension](missing_link): A prototype implementation of the JWT
    authentication.
- [How to create your own authentication strategy](missing_link): Particularly
  for extension developers.

> - [Use [Passport](https://www.npmjs.com/package/passport) strategies](missing

Reviewer 3: fix missing link

Author: @dhmlau sure, I am creating sub-PRs to update each page, and eventually the links will all be updated.

LoopBack's authentication mechanism.

![authentication_overview_request_handle_flow](./imgs/authentication/authentication-overview.png)

As illustrated above, an access token in Authorization header in making request
to the route is handled by the REST server's sequence, which invokes the
authentication action to decode the user profile from token so that controllers
can have the user injected.

For implementing this, all you need to add is the code in programming colors:

- Register the authentication component and JWT extension in your application.
- Enable the authenticate action in sequence.
- Decorate the controller endpoint with `@authenticate()` and inject the user
  passed from the authentication layer.

The rest will be handled by our authentication component
`@loopback/authentication`, which incorporates the authentication mechanism, and
the JWT extension `@loopback/jwt-authentication`, which helps in implementing
JWT-based authentication to the system and should be provided by extension
developers.

The authentication system is highly extensible and pluggable. It's easy to get
started with. While there are more advanced features to explorer - of which are

> beneficial when you build more complicated and larger scale applications. it

Reviewer 3: ```suggestion
beneficial when you build more complicated and larger scale applications, it
```

---
lang: en
title: 'LoopBack 4 Authentication Overview'
keywords: LoopBack 4.0, LoopBack 4, Authentication
sidebar: lb4_sidebar
permalink: /doc/en/lb4/Authentication-overview.html
---

Security is of paramount importance when developing a web or mobile application
and usually consists of two distinct pieces:

- Authentication
- Authorization

Authentication is a process of verifying user/entity to the system, which
enables identified/validated access to the protected routes.

Authorization is a process of deciding if a user can perform an action on a
protected resource.

{% include note.html content=" For a description of an Authorization process, please see [Authorization](Loopback-component-authorization.md). " %}

This document gives you an overview of the authentication system provided in
LoopBack 4.

Let's start with the following scenario: Suppose you want to secure endpoint
`GET /todo` using a
[JWT(JSON web token)](https://github.com/auth0/node-jsonwebtoken#readme)

> strategy. The diagram below shows how such authentication process works with

Reviewer 5: This statement is a bit too technical. We can say we want to limit access to `todo` items to the owner.
